Fix linkjob

ZihanL1 opened the pull request:
1. fix the problem that the first step contains no functional and basis
2. fix the IRC will create two additional_route_parameters (e.g., scf=qc appearing twice)
3. now the link job can do link -j irc
4. change some option-names in cli/irc.py to fix the conflict with other job-type commands
5. update RTD accordingly
6. test the link-job and irc, now they can work well.

Please have a check!

xinglong-zhang reviewed (changes requested):
Some comments for your consideration. There are no tests added for this link irc job, can you add tests in `test_GaussianJobs.py`? @ZihanL1

ZihanL1 commented on chemsmart/cli/gaussian/irc.py:
```diff
@@ -10,62 +11,73 @@
 
 
-@gaussian.command(cls=MyCommand)
+def click_irc_options(f):
```
fixed

ZihanL1 commented on chemsmart/cli/gaussian/irc.py:
```diff
@@ -10,62 +11,73 @@
 
 
-@gaussian.command(cls=MyCommand)
+def click_irc_options(f):
```
fixed

ZihanL1 commented on chemsmart/jobs/gaussian/link.py:
```diff
@@ -71,3 +78,154 @@ def settings_class(cls) -> Type[GaussianLinkJobSettings]:
+        Returns:
+            bool: True if job type is IRC, False otherwise.
+        """
+        return self.settings.job_type == "irc"
```
for link job to determine the `job_type` is` irc `or not, and then use` ircf `and` ircr` to do further process.

ZihanL1 commented on chemsmart/jobs/gaussian/settings.py:
```diff
@@ -1074,37 +1114,50 @@ def _get_route_string_from_jobtype(self):
-        # stable=opt guess=mix
+        # Remove opt keywords (opt, opt=(...), opt=word)
+        pattern = re.compile(
+            r"\bopt\s*(=\s*(\([^)]*\)|\w+))?\s*", re.IGNORECASE
```
fixed

ZihanL1 commented on chemsmart/jobs/gaussian/settings.py:
```diff
@@ -1074,37 +1114,50 @@ def _get_route_string_from_jobtype(self):
+        route_string_final = re.sub(pattern, " ", route_string)
+        # Remove freq keywords for all job types
+        route_string_final = re.sub(
+            r"\bfreq\s*(=\s*\w+)?\s*",
```
added

ZihanL1 commented on docs/source/gaussian_submittransitionstatesearchjobs.rst:
```diff
@@ -145,15 +145,15 @@ IRC-Specific OPTIONS
       -  Compute the Hessian analytically every N predictor steps or every N corrector steps if N<0 (default=6)
 
-   -  -  ``-p, --maxpoints``
+   -  -  ``-mp, --maxpoints``
```
The existing three options may conflict with other options in linkjob. For example, `-c` is also used for `modred -c []`.

ZihanL1 replied:
<img width="639" height="102" alt="image" src="https://github.com/user-attachments/assets/40731d99-2b90-467c-a010-159bd90df2a9" />
If we run **normal irc** and **modred** using option `-c`, they will not conflict, but in **link irc** and **link modred** job, the `-c` will conflict with each other. Can I use one capital letter for irc option?

ZihanL1 commented:
I have added some simple tests and manually tested almost all use cases. 
Additionally, I have introduced the "forward/reverse" option to enable single-direction IRC, which is applicable to both regular IRC and link-IRC.

xinglong-zhang commented on chemsmart/cli/gaussian/gaussian.py:
```diff
@@ -137,6 +137,76 @@ def wrapper_common_options(*args, **kwargs):
+        default=20,
+        help="Step size along reaction path, in units of 0.01 Bohr.",
+    )
+    @click.option(
```
It will better to separte `--forward` and `--reverse` as two click options, as the current convention is a bit confusing.
We can have one click boolean option of `--forward/--no-forward` and another click boolean option of `--reverse/--no-reverse`.
It's strange to put type of boolean and have `--forward/--reverse`.

xinglong-zhang commented on chemsmart/cli/gaussian/gaussian.py:
```diff
@@ -137,6 +137,76 @@ def wrapper_common_options(*args, **kwargs):
+        help="Maximum number of steps along IRC to run.",
+    )
+    @click.option(
+        "-ss",
```
Let's use `-s` here, as it is used as step size in `scan` job.

xinglong-zhang commented on chemsmart/cli/gaussian/gaussian.py:
```diff
@@ -137,6 +137,76 @@ def wrapper_common_options(*args, **kwargs):
+        help="whether to run flat irc or not",
+    )
+    @click.option(
+        "-pt",
```
We can use `-p`.

xinglong-zhang commented on chemsmart/cli/gaussian/gaussian.py:
```diff
@@ -137,6 +137,76 @@ def wrapper_common_options(*args, **kwargs):
+        '["Never", "Always", "Test"].',
+    )
+    @click.option(
+        "-rs",
```
We can just use `-r`.

xinglong-zhang commented on chemsmart/cli/gaussian/gaussian.py:
```diff
@@ -137,6 +137,76 @@ def wrapper_common_options(*args, **kwargs):
+    """Common click options for IRC-related jobs."""
+
+    @click.option(
+        "-fl/",
```
Maybe we can change this to `-F` (same for the non-link irc job).

xinglong-zhang commented on chemsmart/cli/gaussian/gaussian.py:
```diff
@@ -137,6 +137,76 @@ def wrapper_common_options(*args, **kwargs):
+        "LQA, DVV, Euler].",
+    )
+    @click.option(
+        "-rc",
```
We can delete "-rc" to use full option name.

xinglong-zhang commented on chemsmart/cli/gaussian/gaussian.py:
```diff
@@ -137,6 +137,76 @@ def wrapper_common_options(*args, **kwargs):
+        "|N| corrector steps if N<0. ",
+    )
+    @click.option(
+        "-mp",
```
can just use `-p`.

xinglong-zhang commented on chemsmart/cli/gaussian/gaussian.py:
```diff
@@ -137,6 +137,76 @@ def wrapper_common_options(*args, **kwargs):
+        help="Number of points along reaction path to examine.",
+    )
+    @click.option(
+        "-mc",
```
can just use `-c`.

xinglong-zhang reviewed (changes requested):
Some coments for your consideration, @ZihanL1.

ZihanL1 commented on tests/test_repatterns.py:
```diff
@@ -5,9 +5,122 @@
+    ), "Should match: freq = analytical"
+    # Test that it doesn't match partial words
+    assert (
+        pattern.search("# b3lyp opt frequency") is not None
```
Chemsmart will automatically use 'freq' instead of other variants, such as 'frequency', even if 'frequency' is also recognizable by Gaussian. 
This test is unnecessary and has been deleted.
